### ISSUE TITLE
Syntax-based covariance detection can misfire

### ISSUE THREAD

**Issue by jonas-schievink** [joshua-maros/ouroboros#19]
I see that the macro looks at the name of the field types to determine whether the type is covariant, but users can define their own invariant types with the same name:

```rust
struct Box<T> {
    inner: RefCell<T>,
}
```

It looks like the current check would treat this type as covariant.

**Comment by joshua-maros** [comment]
This is a good point. I am not sure what the best way to proceed is, so input is welcome. The simplest solution would be to require manually specifying whether or not each field is covariant, but this seems like a lot of work and was the situation I was hoping to avoid with the automatic detection in the first place. One thought I had which I have not yet investigated would be having the automatic detection automatically replace any time it sees things like `Box` with their fully qualified name like `::std::boxed::Box`, however, this might be unintuitive and annoying to debug if the user is genuinely trying to use a custom type.

**Comment by joshua-maros** [comment]
Unrelated, just briefly wanted to say thank you for working on Rust-Analyzer, it's a super amazing tool that I love working with.

**Comment by jonas-schievink** [comment]
Hmm, one possibility instead of outright *replacing* the type with the "right one" would be to emit an assertion that will just cause a type mismatch when the user attempted to override the type. This seems less subtle than replacing the type, but the error message will probably not be great either.

**Comment by joshua-maros** [comment]
@jonas-schievink I've experimented with your suggestion and found an approach that I think generates acceptable error messages:
```
error[E0599]: no function or associated item named `is_std_box_type` found for struct 
`CheckIfTypeIsStd<Box<String>>` in the current scope
  --> $DIR/refuse_non_std_box.rs:21:1
   |
21 | #[self_referencing]
   | ^^^^^^^^^^^^^^^^^^^ function or associated item not found in `CheckIfTypeIsStd<Box<String>>`
   |
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
```
I've added a test to make sure this error is generated when using a non-std box type and published the changes as crate version `0.9.0`. Any comments or feedback are appreciated.
